File: stacks/go/checks/architecture_test.go
// Structural fitness test. Fails if module-boundary invariants from
// docs/architecture.md are violated. Runs as part of `go test ./...`.
package checks

import (
	"go/parser"
	"go/token"
	"io/fs"
	"path/filepath"
	"strings"
	"testing"
)

var allowed = map[string]map[string]bool{
	"core":      {"core": true},
	"adapters":  {"core": true, "adapters": true},
	"app":       {"core": true, "adapters": true, "app": true},
	"interface": {"app": true, "interface": true},
}

func layerOf(path string) string {
	for _, layer := range []string{"core", "adapters", "app", "interface"} {
		if strings.Contains(path, "/"+layer+"/") || strings.HasSuffix(path, "/"+layer) {
			return layer
		}
	}
	return ""
}

func sourceFiles(t *testing.T) []string {
	t.Helper()
	var files []string
	err := filepath.WalkDir("../src", func(path string, d fs.DirEntry, err error) error {
		if err != nil {
			return err
		}
		if !d.IsDir() && strings.HasSuffix(path, ".go") && !strings.HasSuffix(path, "_test.go") {
			files = append(files, path)
		}
		return nil
	})
	if err != nil {
		t.Fatalf("walk: %v", err)
	}
	if len(files) == 0 {
		t.Fatal("no source files found under ../src — architecture test would silently pass")
	}
	return files
}

func checkFileImports(t *testing.T, file string) {
	t.Helper()
	layer := layerOf(file)
	if layer == "" {
		return
	}
	fset := token.NewFileSet()
	parsed, err := parser.ParseFile(fset, file, nil, parser.ImportsOnly)
	if err != nil {
		t.Fatalf("%s: %v", file, err)
	}
	for _, imp := range parsed.Imports {
		path := strings.Trim(imp.Path.Value, `"`)
		target := layerOf("/" + path + "/")
		if target == "" {
			continue
		}
		if !allowed[layer][target] {
			t.Errorf("%s (%s) imports %s (%s): violates docs/architecture.md",
				file, layer, path, target)
		}
	}
}

func TestArchitectureBoundaries(t *testing.T) {
	for _, f := range sourceFiles(t) {
		checkFileImports(t, f)
	}
}

File: stacks/typescript/checks/architecture.test.ts
/**
 * Structural fitness test. Fails if module-boundary invariants from
 * docs/architecture.md are violated. Run as part of `npm test`.
 */
import { describe, expect, it } from "vitest";
import { readFileSync } from "node:fs";
import { globSync } from "node:fs";
import { join } from "node:path";

type Layer = "core" | "adapters" | "app" | "interface";

const allowed: Record<Layer, Layer[]> = {
  core: ["core"],
  adapters: ["core", "adapters"],
  app: ["core", "adapters", "app"],
  interface: ["app", "interface"],
};

function layerOf(path: string): Layer | null {
  const m = path.match(/src\/(core|adapters|app|interface)\//);
  return (m?.[1] as Layer) ?? null;
}

describe("architecture invariants", () => {
  const files = globSync("src/**/*.{ts,tsx}", { cwd: process.cwd() });
  for (const file of files) {
    const layer = layerOf(file);
    if (!layer) continue;
    it(`${file} only imports allowed layers`, () => {
      const src = readFileSync(join(process.cwd(), file), "utf8");
      const imports = [...src.matchAll(/from\s+["']([^"']+)["']/g)].map(
        (m) => m[1]!,
      );
      for (const spec of imports) {
        const target = spec.match(/@(core|adapters|app|interface)\b/);
        if (!target) continue;
        const t = target[1] as Layer;
        expect(
          allowed[layer].includes(t),
          `${file} (${layer}) → ${spec} (${t}) violates docs/architecture.md`,
        ).toBe(true);
      }
    });
  }
});
